### PR TITLE
fix(parser): remove extra parenthesis when displaying nested operations

### LINKED ISSUE #7635: bug: roundtrip parsing for `OR` does not return same expression
### Describe the bug

**Not too sure if this is a bug.**

- What should be the default behaviour of `fmt` for a query. Should it return the same query?
- Why do we need to wrap complex expressions in parenthesis? Is this just for debugging?

```
---- run_sqlsmith_on_frontend_109 ----
Roundtrip test failed
Start: SELECT TRUE OR TRUE IS NOT FALSE
Roundtrip: SELECT true OR (true IS NOT FALSE)
Start AST: Query(Query { with: None, body: Select(Select { distinct: All, projection: [UnnamedExpr(BinaryOp { left: Value(Boolean(true)), op: Or, right: IsNotFalse(Value(Boolean(true))) })], from: [], lateral_views: [], selection: None, group_by: [], having: None }), order_by: [], limit: None, offset: None, fetch: None })
Roundtrip AST: Query(Query { with: None, body: Select(Select { distinct: All, projection: [UnnamedExpr(BinaryOp { left: Value(Boolean(true)), op: Or, right: Nested(IsNotFalse(Value(Boolean(true)))) })], from: [], lateral_views: [], selection: None, group_by: [], having: None }), order_by: [], limit: None, offset: None, fetch: None })
```

I guess it is due to https://github.com/risingwavelabs/risingwave/blob/5938e35c02c2091d165b9fb70107d9399f86a0e0/src/sqlparser/src/ast/mod.rs#L562-L572.

### To Reproduce

_No response_

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by kwannoel**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Only display parenthesis if function requires it, or if user explicitly supplied them:
e.g.
```sql
select max(v1);
select (1 + (2 - 1));
```

This should unparse to original:
```sql
select NOT a IS NULL; -- unparse to itself
select NOT (a IS NULL); -- above should not unparse to this
```

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)

Closes #7635

### CODE REVIEW

**Comment by kwannoel** [comment]
Realized the `fmt` implementation for `Expr` is used to test for precedence.
Better way to check is by directly looking at the AST output. Requires quite a few changes to test, work on it again when I have some bandwidth.

File: src/sqlparser/src/ast/mod.rs [hunk 1]
```diff
@@ -557,24 +551,6 @@ impl fmt::Display for Expr {
     }
 }
 
-/// Wrap complex expressions with necessary parentheses.
-/// For example, `a > b LIKE c` becomes `a > (b LIKE c)`.
-fn fmt_expr_with_paren(e: &Expr) -> String {
```
Comment (BugenZhao) [inline src/sqlparser/src/ast/mod.rs:562]: I'm also curious why the upstream adds this extra step. 🤔 Is it just for better readability after formatting?

**Review by BugenZhao — APPROVED** [review]

**Comment by kwannoel** [comment]
Fails at
```
select CAST((SMALLINT '240') >> (SMALLINT '27901') & 1  * (SMALLINT '18731') AS BOOLEAN);
```

Probably #7641 

Previously we will generate brackets due to `fmt`. That's why it is rare to encounter.

**Comment by kwannoel** [comment]
Let @xiangjinwu take a look first as he had some comments: https://github.com/risingwavelabs/risingwave/issues/7671#issuecomment-1414983049.

**Review by xiangjinwu — APPROVED** [review]

**Pull request merged**